dhcpv6-ia: allow up to 64 bit wide hostid

mikma opened the pull request:
This PR implements up to 64 bit wide hostid for dynamic and static leases.

dedeckeh commented on src/odhcpd.h:
```diff
@@ -199,7 +199,7 @@ struct dhcp_assignment {
 
 	uint32_t addr;
-	uint32_t assigned;
+	uint64_t assigned;
```
Changing assigned into an uint64_t makes the code hard to understand and read in case of the prefix delegation logic as assigned is used to assign the subnet id

mikma replied:
I have changed `assigned` into an anonymous union with members `uint64_t assigned_host_id` and `uint32_t assigned_subnet_id`.

dedeckeh commented on src/dhcpv6-ia.c:
```diff
@@ -688,12 +689,32 @@ static bool assign_na(struct interface *iface, struct dhcp_assignment *a)
+				mask_high = UINT32_MAX;
+			else
+				mask_high = (1 << (iface->dhcpv6_hostid_len - 32)) - 1;
+			do {
```
Align with the existing coding style by inserting an empty line before the do statement

mikma replied:
Fixed.

mikma commented on src/config.c:
```diff
@@ -754,6 +757,17 @@ int config_parse_interface(void *data, size_t len, const char *name, bool overwr
+	if ((c = tb[IFACE_ATTR_DHCPV6_HOSTID_LEN])) {
+		uint32_t hostid_len = blobmsg_get_u32(c);
+
+		if (hostid_len >= 12 && hostid_len <= 64)
```
Fixed.

dedeckeh commented:
Can you squash all the commits into one commit as they're all related to the same feature

mikma commented:
I have squashed and rebased on master at git.openwrt.org.

I also added a check against the reserved IPv6 interface identifier ranges listed at http://www.iana.org/assignments/ipv6-interface-ids .

dedeckeh commented:
This fails to compile for me with the following error

src/dhcpv6-ia.c: In function 'dhcpv6_ia_write_statefile':
src/dhcpv6-ia.c:367:54: error: format '%zx' expects argument of type 'size_t', but argument 10 has type 'uint64_t' {aka 'long long unsigned int'} [-Werror=format=]
      ctxt.buf_idx = snprintf(ctxt.buf, ctxt.buf_len, "# %s %s %x %s%s %"PRId64" %zx %u ",
                                                      ^~~~~~~~~~~~~~~~~~~
src/dhcpv6-ia.c:374:9:
         (ctxt.c->flags & OAF_DHCPV6_NA) ? ctxt.c->assigned_host_id : ctxt.c->assigned_subnet_id,
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/dhcpv6-ia.c:367:83: note: format string is defined here
      ctxt.buf_idx = snprintf(ctxt.buf, ctxt.buf_len, "# %s %s %x %s%s %"PRId64" %zx %u ",
                                                                                 ~~^
                                                                                 %llx

As assigned_hostìd is an uint64_t the best option is to move this to two separate snprintf lines depending on the assignment type

mikma commented:
I have duplicated the snprintf statements and I use the PRIx64/32-macros in the format string now.

dedeckeh commented:
Patch pushed to master with a minor modification (https://git.openwrt.org/?p=project/odhcpd.git;a=commit;h=16667699228101fda936068e3d11c0cf19270e34); thx
